clang/lib/CIR/Dialect/Transforms/FlattenCFG.cpp

```diff
@@ -869,11 +869,41 @@ class CIRTernaryOpFlattening
   }
 };
 
+// Hoist all static allocas to the entry block.
+class CIRAllocaOpFlattening : public mlir::OpRewritePattern<mlir::cir::AllocaOp> {
+public:
+  using OpRewritePattern<mlir::cir::AllocaOp>::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::AllocaOp allocaOp,
+                  mlir::PatternRewriter &rewriter) const override {
+    mlir::cir::FuncOp funcOp = allocaOp->getParentOfType<mlir::cir::FuncOp>();
+    if (!funcOp || funcOp.getRegion().empty())
+      return mlir::failure();
+
+    mlir::Block &entryBlock = funcOp.getRegion().front();
+
+    if (allocaOp->getBlock() == &entryBlock)
+      return mlir::failure();
+
+    // Don't hoist allocas with dynamic alloca size.
+    if (allocaOp.getDynAllocSize() != mlir::Value())
+      return mlir::failure();
+
+    // It is cheaper to call `mlir::Operation::moveBefore` than using rewriter.
+    // So we prefer to manually here.
+    mlir::Operation *insertPoint = &*entryBlock.begin();
+    allocaOp->moveBefore(insertPoint);
```

Comment:
In a rewrite pattern, all IR mutations **must** be performed through the rewriter, or it may crash. See https://mlir.llvm.org/docs/PatternRewriter/#restrictions :

> Within the rewrite section of a pattern, the following constraints apply:
> 
> - All IR mutations, including creation, must be performed by the given PatternRewriter. This class provides hooks for performing all of the possible mutations that may take place within a pattern. For example, this means that an operation should not be erased via its erase method. To erase an operation, the appropriate PatternRewriter hook (in this case eraseOp) should be used instead.
> - [...]

Reply:
https://mlir.llvm.org/getting_started/Debugging/#detecting-invalid-api-usage mentions `-DMLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS=ON`, which might be useful for development? I haven't tried it though.

Reply:
I wished we had a clang format thing for preventing us to use this idiom in rewriters.

Reply:
> https://mlir.llvm.org/getting_started/Debugging/#detecting-invalid-api-usage mentions `-DMLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS=ON`, which might be useful for development? I haven't tried it though.

Maybe we can enable this for github testing, so we catch those in PRs. Now that Nathan disabled unrelated tests perhaps it won't feel like it's taking too much longer.